File: Problem1.java
// Time Complexity : O(1)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :
class SnakeGame {
    LinkedList<int[]> snake;
    int h; int w;
    boolean[][] visited;
    int[] head;
    int[][] foodList;
    int i;

    public SnakeGame(int width, int height, int[][] food) {
        this.h = height;
        this.w = width;
        this.snake = new LinkedList<>();
        this.visited = new boolean[h][w];
        this.head = new int[]{0,0};
        this.snake.addFirst(new int[]{head[0],head[1]});
        this.foodList = food;
    }
    
    public int move(String direction) {
        if(direction.equals("U"))
            head[0]--;
        else if(direction.equals("D"))
            head[0]++;
        else if(direction.equals("R"))
            head[1]++;
        else 
            head[1]--;
        if(head[0] < 0 || head[0] >= h || head[1] < 0|| head[1] >=w||visited[head[0]][head[1]])
            return -1;
        if(i < foodList.length){
            if(head[0] == foodList[i][0] && head[1] == foodList[i][1]){
                snake.addFirst(new int[]{head[0],head[1]});
                visited[head[0]][head[1]] = true;
                i++;
                return snake.size()-1;
            }
        }
        snake.addFirst(new int[]{head[0],head[1]});
        visited[head[0]][head[1]] = true;
        snake.pollLast();
        int[] curr = snake.peekLast();
        visited[curr[0]][curr[1]] = false;
        return snake.size()-1;
        
        
    }
}

File: problem1.cpp
/*
Time complexity: O(1)
Space complexity: O(n)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
*/

struct Node {
  int key;
  int value;
  int freq;
  list<int>::const_iterator it;
};

class LFUCache {
 public:
  LFUCache(int capacity) : capacity(capacity), minFreq(0) {}

  int get(int key) {
    if (!keyToNode.count(key))
      return -1;

    Node& node = keyToNode[key];
    touch(node);
    return node.value;
  }

  void put(int key, int value) {
    if (capacity == 0)
      return;
    if (keyToNode.count(key)) {
      Node& node = keyToNode[key];
      node.value = value;
      touch(node);
      return;
    }

    if (keyToNode.size() == capacity) {
      const int keyToEvict = freqToList[minFreq].back();
      freqToList[minFreq].pop_back();
      keyToNode.erase(keyToEvict);
    }

    minFreq = 1;
    freqToList[1].push_front(key);
    keyToNode[key] = {key, value, 1, cbegin(freqToList[1])};
  }

 private:
  int capacity;
  int minFreq;
  unordered_map<int, Node> keyToNode;
  unordered_map<int, list<int>> freqToList;

  void touch(Node& node) {
    const int prevFreq = node.freq;
    const int newFreq = ++node.freq;
    freqToList[prevFreq].erase(node.it);
    if (freqToList[prevFreq].empty()) {
      freqToList.erase(prevFreq);
      if (prevFreq == minFreq)
        ++minFreq;
    }

    freqToList[newFreq].push_front(node.key);
    node.it = cbegin(freqToList[newFreq]);
  }
};